app/code/Magento/InventoryGroupedProduct/Plugin/OutOfStockFilter.php

```diff
@@ -0,0 +1,84 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryGroupedProduct\Plugin;
+
+use Magento\Framework\DataObject;
+use Magento\Catalog\Model\Product;
+use Magento\GroupedProduct\Model\Product\Type\Grouped;
+use Magento\Store\Model\StoreManagerInterface;
+use Magento\InventorySalesApi\Api\IsProductSalableInterface;
+use Magento\InventorySalesApi\Api\StockResolverInterface;
+use Magento\InventorySalesApi\Api\Data\SalesChannelInterface;
+
+/**
+ * Removes out of stock products from cart candidates when appropriate.
+ */
+class OutOfStockFilter
+{
+    /**
+     * @var StoreManagerInterface
+     */
+    private $storeManager;
+
+    /**
+     * @var IsProductSalableInterface
+     */
+    private $isProductSalable;
+
+    /**
+     * @var StockResolverInterface
+     */
+    private $stockResolver;
+
+    /**
+     * @param StoreManagerInterface $storeManager
+     * @param IsProductSalableInterface $isProductSalable
+     * @param StockResolverInterface $stockResolver
+     */
+    public function __construct(
+        StoreManagerInterface $storeManager,
+        IsProductSalableInterface $isProductSalable,
+        StockResolverInterface $stockResolver
+    ) {
+        $this->storeManager = $storeManager;
+        $this->isProductSalable = $isProductSalable;
+        $this->stockResolver = $stockResolver;
+    }
+
+    /**
+     * Removes out of stock products for requests that don't specify the super group.
+     *
+     * @param Grouped $subject
+     * @param array|string $result
+     * @param \Magento\Framework\DataObject $buyRequest
+     * @return string|array
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function afterPrepareForCartAdvanced(Grouped $subject, $result, DataObject $buyRequest)
+    {
+        if (!is_array($result) && $result instanceof Product) {
+            $result = [$result];
+        }
+
+        // Only remove out-of-stock products if no quantities were specified
+        if (is_array($result) && !empty($result) && !$buyRequest->getData('super_group')) {
+            $websiteCode = $this->storeManager->getWebsite()
+                ->getCode();
+            $stockId = $this->stockResolver->execute(SalesChannelInterface::TYPE_WEBSITE, $websiteCode)
+                ->getStockId();
+
+            foreach ($result as $index => $cartItem) {
+                if (!$this->isProductSalable->execute($cartItem->getSku(), $stockId)) {
```

Comment:
While this solution is functional, there are still at least 2 alternatives to duplicating the logic in this plugin:
* Longer solution would be to adapt `StockStatusCriteriaInterface` and `StockStatusRepositoryInterface::getList` to take advantage of MSI. This might require building additional plugins and and manipulation with queries in the Inventory modules. This solution, while could be functional, does not have too much sense, as all of this work would be done to extend a deprecated API
* Simpler and more clean solution would be refactoring original plugin (`Magento\GroupedCatalogInventory\Plugin\OutOfStockFilter`) to use `StockRegistryInterface` which is already adapted by MSI to support multi-stock capabilities.

Reply:
Thank you for your consideration. I have reworked this solution on CE's OutOfStockFilter plugin.
![Screenshot from 2019-07-12 16-28-21](https://user-images.githubusercontent.com/37332845/61131626-5a78fd00-a4c2-11e9-997f-351fe678ebab.png)
, where $this->stockRegistry is an object of StockRegistryInterface class.

Can it be a right solution?